src/harness/core/collections.ts

```diff
@@ -0,0 +1,542 @@
+            }
+        }
+
+        public * keys() {
+            const keys = this._keys;
+            const indices = this.getIterationOrder();
+            const version = this._version;
+            this._copyOnWrite = true;
+            try {
+                if (indices) {
+                    for (const i of indices) {
+                        yield keys[i];
+                    }
+                }
+                else {
+                    yield* keys;
+                }
+            }
+            finally {
+                if (version === this._version) {
+                    this._copyOnWrite = false;
+                }
+            }
+        }
+
+        public * values() {
+            const values = this._values;
+            const indices = this.getIterationOrder();
+            const version = this._version;
+            this._copyOnWrite = true;
+            try {
+                if (indices) {
+                    for (const i of indices) {
+                        yield values[i];
+                    }
+                }
+                else {
+                    yield* values;
+                }
+            }
+            finally {
+                if (version === this._version) {
+                    this._copyOnWrite = false;
+                }
+            }
+        }
+
+        public * entries() {
+            const keys = this._keys;
+            const values = this._values;
+            const indices = this.getIterationOrder();
+            const version = this._version;
+            this._copyOnWrite = true;
+            try {
+                if (indices) {
+                    for (const i of indices) {
+                        yield [keys[i], values[i]] as [K, V];
+                    }
+                }
+                else {
+                    for (let i = 0; i < keys.length; i++) {
+                        yield [keys[i], values[i]] as [K, V];
+                    }
+                }
+            }
+            finally {
+                if (version === this._version) {
+                    this._copyOnWrite = false;
+                }
+            }
+        }
+
+        public [Symbol.iterator]() {
+            return this.entries();
+        }
+
+        private writePreamble() {
+            if (this._copyOnWrite) {
+                this._keys = this._keys.slice();
+                this._values = this._values.slice();
+                if (this._order) this._order = this._order.slice();
+                this._copyOnWrite = false;
+            }
+        }
+
+        private writePostScript() {
+            this._version++;
+        }
+
+        private getIterationOrder() {
+            if (this._order) {
+                const order = this._order;
+                return this._order
+                    .map((_, i) => i)
+                    .sort((x, y) => order[x] - order[y]);
+            }
+            return undefined;
+        }
+    }
+
+    export class SortedSet<T> {

```

Comment:
i am not seeing this used anywhere.. is it needed?

Reply:
It was originally used as part of the file watcher functionality.